### ISSUE TITLE
`cast_possible_truncation` triggers when the value is guranteed to not be truncated

### ISSUE THREAD

**Issue by Luro02** [rust-lang/rust-clippy#7486]
Lint name: `cast_possible_truncation`

I tried this code:

```rust
#![deny(clippy::cast_possible_truncation)]

fn modulo(number: u64, other: u32) -> u32 {
    // a u64 with guranteed value to be less than or equal to u32::max_value()
    let other_u64 = other as u64;
    // modulo gurantees that the following invariant holds:
    // result < other_u64
    // which implies that result < u32::max_value()
    let result = number % other_u64;
    result as u32
}

fn main() {
    assert_eq!(modulo(11_u64, 5_u32), 1_u32);
}
```

I expected to see this happen: the lint should not trigger

Instead, this happened: the lint about a possible truncation has been emitted

### Meta

- `cargo clippy -V`: `clippy 0.1.55 (02718709 2021-07-22)`
- `rustc -Vv`:
```
rustc 1.55.0-nightly (027187094 2021-07-22)
binary: rustc
commit-hash: 027187094ee05011d6602f5742f550851ccc7fd6
commit-date: 2021-07-22
host: x86_64-pc-windows-msvc
release: 1.55.0-nightly
LLVM version: 12.0.1
```

**Comment by dtolnay** [comment]
I've hit an even simpler form of this, where the modulus is an integer literal. I believe that Clippy should be able to avoid triggering in the literal case at least.

```rust
pub fn f(val: u64) {
    let digit = (val % 62) as u8;
    println!("{}", digit);
}
```

```console
error: casting `u64` to `u8` may truncate the value
 --> src/main.rs:2:17
  |
2 |     let digit = (val % 62) as u8;
  |                 ^^^^^^^^^^^^^^^^
  |
  = note: `-D clippy::cast-possible-truncation` implied by `-D clippy::pedantic`
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#cast_possible_truncation
```

**Comment by jyn514** [comment]
Hmm, #7666 must not have been complete because this still lints on things like
```rust
pub fn foo(i: u16) -> u8 {
    (i % 256) as u8
}
```

**Comment by rdrpenguin04** [comment]
I know why that is. Lemme fork and PR quick.

There's basically a math error going on: `n % m` has a maximum value of `m - 1`, not `m` as the current `apply_reductions` function calculates.
